Privacy related permission

yl33 opened the pull request:
### Change Description ###

- PPM APIs launch Popup UI to get user response for 'permission request' for privacy related privileges. Originally the UI was drawn by daemon but we separated UI part from the daemon and made daemon to launch popup app since Tizen 6.0(and we decide to apply this change to 5.5). Added 'note' for those changes.
- Modify dotnet PPM API document file's name like native and web.
- TODO: This change need to be applied to 5.5 doc.

mijin85cho commented:
Hello, @yl33 
In the description, it is said that this update needs to be applied to Tizen 5.5 docs.

In tizen-docs repository,
 - **master** branch is to handle Tizen 5.5 (Currently latest) docs, 
 - **tizen_6.0_prepare** branch is to handle Tizen 6.0 docs

So the updates on tizen_6.0_prepare will be applied on docs.tizen.org when Tizen 6.0 is published.
Don't we need to use master branch according to the description?

Side question : So does it mean that this update is officially applied since Tizen 6.0, but Tizen 5.5 is also supported? What is different with "applied since Tizen 5.5"?

yl33 commented:
Hi, @mijin85cho 
We made this change in 6.0 but it'll be applied to 5.5, too. But released Tizen 5.5-M1 and M2- doesn't include this changes(It's one of the reasons why I didn't raise PR to master branch). We're applying the changes to both tizen and tizen_5.5 branch and syncing the changes to SPIN(and is applied to 5.5 product) but don't know whether it is okay to apply the changes to the latest Tizen doc. 
If applying this changes to the latest Tizen doc is okay then I'll raise PR for master branch when review of this change on this branch is done. If modification is needed, I don't want to make redundant updates for both PR. Or.. is it applied to tizen_6.0_prepare branch if this change is applied to master branch?

mijin85cho commented:
Hello, @yl33 

>If applying this changes to the latest Tizen doc is okay then I'll raise PR for master branch when review of this change on this branch is done. If modification is needed, I don't want to make redundant updates for both PR. Or.. is it applied to tizen_6.0_prepare branch if this change is applied to master branch?

Actually tizen_6.0_prepare stores the **additional updates for Tizen 6.0** comparing to Tizen 5.5 .

So what I meant with _"the updates on tizen_6.0_prepare will be applied on docs.tizen.org when Tizen 6.0 is published._" is :

When Tizen 6.0 is published, 
the fixes on tizen_6.0_prepare will be **added** to master branch,
and from that time we will consider master branch as the branch to handle Tizen 6.0.

You can consider master branch as the base. 
If master branch is updated, the fix will be kept also when Tizen 6.0 is published.
So if master branch is updated, you don't need to update same fix to tizen_6.0_prepare.

**Note :** If master branch is updated, the updates will be shown on docs.tizen.org later.

yl33 commented:
Hello, @mijin85cho 
Thank you for the kind explanation.

This change is not included in Tizen 5.5 M1 and M2 so
I think applying this change to **master** branch at this moment can misguide developers.

If **Tizen 5.5 M3** or **Product with Platform version 5.5** is released then it should be updated to **master** branch.
But I guess there's no branch for **5.5 M3 release** currently and..
I don't know whether **5.5 M3 release** or **product release** will be done before **6.0 release**. 
(If **6.0 release** is done before **5.5 M3 release** then 
**tizen_6.0_release** branch's changes would be applied to **master** branch.)

Therefore, I think **tizen_6.0_prepare** branch would be the correct one at this moment.

mijin85cho commented:
Hello, @yl33 

Thanks for your explanation.

With your explanation, and the same reason that you explained, 
I agree with you that the tizen_6.0_prepare is the right branch to proceed.

When review is done, I'll merge this PR, targeting tizen_6.0_prepare branch.

mijin85cho commented on docs/application/dotnet/guides/security/privacy-related-permissions.md:
```diff
@@ -17,6 +17,12 @@ The main features of the `Tizen.Security.PrivacyPrivilegeManager` class include:
+>
+> Popups by `ppm_request_permission()` and `ppm_request_permissions()` are launched as 'group mode' with caller apps since Tizen 5.5.   
+> If the popup is terminated without full responses, all remained requests that aren't responded by the user will be treated as if the user selected `PRIVACY_PRIVILEGE_MANAGER_REQUEST_RESULT_DENY_ONCE` and you can request the permission again anytime you want.  
+> If you need to request multiple privileges then use `ppm_request_permissions()`.
```
Another opinion to the last comment.

>Line 24:
>Slightly changed the structure keeping the original sentence intact.
>
>>In addition, user can request for multiple privileges using `ppm_request_permissions()`

In the note, `ppm_request_permissions()` is mentioned from the 1st line.
If `ppm_request_permissions()` is describe 1st time at this line, than "In addition" would be acceptable, as it has the nuance of describing additional option.

But as `ppm_request_permissions()` is already mentioned with `ppm_request_permission()`,
I think "in addition" here is not natural. This line is describing which one should be used among 2 options already presented to the reader.

| Original | Suggested |
| ---- | ---- |
| There is AA and AAs.<br>...<br>If you need to use plural, use AAs | There is AA and AAs.<br>...<br>In addition, user can use plural using AAs |

Another point is that the function name itself gives a hint to which one should be used for which case. Because of it, I'm not sure the effect of "In addition".

`ppm_request_permissions()` would be used for multiple privileges, rather then `ppm_request_permission()`.

So how about not using "In addition"? I'm not sure with that update is quite effective comparing to the original.
Please give your opinion.

yl33 replied:
Modified the note a little bit. Please check the modified one.

nawab-samsung replied:
> 
> 
> Another opinion to the last comment.
> 
> > Line 24:
> > Slightly changed the structure keeping the original sentence intact.
> > > In addition, user can request for multiple privileges using `ppm_request_permissions()`
> 
> In the note, `ppm_request_permissions()` is mentioned from the 1st line.
> If `ppm_request_permissions()` is describe 1st time at this line, than "In addition" would be acceptable, as it has the nuance of describing additional option.
> 
> But as `ppm_request_permissions()` is already mentioned with `ppm_request_permission()`,
> I think "in addition" here is not natural. This line is describing which one should be used among 2 options already presented to the reader.
> Original 	Suggested
> There is AA and AAs.
> ...
> If you need to use plural, use AAs 	There is AA and AAs.
> ...
> In addition, user can use plural using AAs
> 
> Another point is that the function name itself gives a hint to which one should be used for which case. Because of it, I'm not sure the effect of "In addition".
> 
> `ppm_request_permissions()` would be used for multiple privileges, rather then `ppm_request_permission()`.
> 
> So how about not using "In addition"? I'm not sure with that update is quite effective comparing to the original.
> Please give your opinion.

@mijin85cho 
- In addition, is used for saying that something extra exists or is happening together with the thing that we are talking about. 

- FunctionX() can do this and in addition, Function() can do that also.
  I added **in addition** following this. 
- Also, function names are not always written keeping context in mind and I think irrespective of whether function name is singular or plural, the supporting text must be considered and written as per the situation. 
- What I intend to say is if I go by your opinion about the function `ppm_request_permissions()`, this function as per the name must always support be used for requesting multiple permission**s**, which is not true. 
- We definitely can also request for single permission that's why the author had to make it clear that we can request more than one permission too.

So I would like to know your opinion on this If you don't agree. We can go with:

The user can also request multiple privileges using ppm_request_permissions().

Can you please suggest a better rephrase, if you don't like the above rephrases.

mijin85cho replied:
@nawab-samsung I still don't think the line is for something extra or happening together. To me, the line is just describing one of two functions. Perhaps that makes me to think "In addition" or "Also" too wordy for the simple sentence. 
Also I don't know why should we spend so much energy to this sentence while other issues to be checked are left.

> In addition, is used for saying that something extra exists or is happening together with the thing that we are talking about.

As the developer guide, I think the author's suggestion is not bad. How about keeping it?

yl33 commented on docs/application/dotnet/guides/security/privacy-related-permissions.md:
```diff
@@ -17,6 +17,12 @@ The main features of the `Tizen.Security.PrivacyPrivilegeManager` class include:
 
 For a list of privacy-related privileges, see [Security and API Privileges](../../tutorials/sec-privileges.md).
 
+> **Note**
```
Done

yl33 commented on docs/application/dotnet/guides/security/overview.md:
```diff
@@ -13,7 +13,7 @@ You can use the following security features in your .NET applications:
 
     You can retrieve information about existing permissions granted by a privilege. The information can be delivered to the user as a notification.
 
--   [Privacy-related Permissions](requesting-permissions.md)
```
Done

yl33 commented on docs/application/dotnet/tutorials/sec-privileges.md:
```diff
@@ -37,7 +37,7 @@ switched it **on**.
 > **Note**  
```
Done

yl33 replied:
Done

yl33 commented on docs/application/native/guides/security/privacy-related-permissions.md:
```diff
@@ -17,6 +17,12 @@ The main features of the Privacy Privilege Manager API include:
+>
+> Popups by `ppm_request_permission()` and `ppm_request_permissions()` are launched as 'group mode' with caller apps since Tizen 5.5.   
+> If the popup is terminated without full responses, all remained requests that aren't responded by the user will be treated as if the user selected `PRIVACY_PRIVILEGE_MANAGER_REQUEST_RESULT_DENY_ONCE` and you can request the permission again anytime you want.  
+> If you need to request multiple privileges then use `ppm_request_permissions()`.
```
Modified the note a little bit. Please check the modified one.

yl33 commented on docs/application/web/guides/security/privacy-related-permissions.md:
```diff
@@ -18,6 +18,11 @@ The main features of the Privacy Privilege API include:
+>
+> Popups by `ppm_request_permission()` and `ppm_request_permissions()` are launched as 'group mode' with caller apps since Tizen 5.5.   
+> If the popup is terminated without full responses, all remained requests that aren't responded by the user will be treated as if the user selected `PRIVACY_PRIVILEGE_MANAGER_REQUEST_RESULT_DENY_ONCE` and you can request the permission again anytime you want.  
+> If you need to request multiple privileges then use `ppm_request_permissions()`.
```
Modified the note a little bit. Please check the modified one.

nawab-samsung commented on docs/application/dotnet/guides/security/privacy-related-permissions.md:
```diff
@@ -17,6 +17,12 @@ The main features of the `Tizen.Security.PrivacyPrivilegeManager` class include:
 
 For a list of privacy-related privileges, see [Security and API Privileges](../../tutorials/sec-privileges.md).
 
+> [!NOTE]
```
@yl33  Kindly refer to our [custom style guide](https://github.com/Samsung/tizen-docs/blob/master/styleguide/custom-style.md) to know more about new note styling.

We do not need an empty space after note tag, so:
> [!NOTE]
> Since Tizen 5.5, pop-ups by `RequestPermission()` and `RequestPermissions()` are launched as `group mode` with the caller apps.   
> If the pop-up is terminated without full response, all the remaining requests that aren't responded by the user will be interpreted as 'Deny' action on behalf of the user. In this case, the app gets `RequestResult.DenyOnce` response to those permissions. The app can again request for those **automatically denied** permissions anytime it needs.  
> User can use `RequestPermissions()` to request for multiple privileges instead of calling `RequestPermission()` multiple times.


@yl33 I slightly modified the sentence for more clarity, request you to verify the sentence technically and if you feel this is fine, kindly incorporate the changes

yl33 replied:
> User can use `RequestPermissions()` to request for multiple privileges instead of calling `RequestPermission()` multiple times.
This note is not for the users but for the developers. Modified your comment slightly.
Modified other phrases as you mentioned.

nawab-samsung replied:
@yl33 thanks for verifying my changes and clarifying the use-case for the note.

nawab-samsung commented on docs/application/native/guides/security/privacy-related-permissions.md:
```diff
@@ -17,6 +17,12 @@ The main features of the Privacy Privilege Manager API include:
 
+> [!NOTE]
+>
+> Since Tizen 5.5, pop-ups by `ppm_request_permission()` and `ppm_request_permissions()` are launched as `group mode` with the caller apps.   
```
@yl33  Kindly refer to our [custom style guide](https://github.com/Samsung/tizen-docs/blob/master/styleguide/custom-style.md) to know more about new note styling.

We do not need an empty space after note tag, so:
> [!NOTE]
> Since Tizen 5.5, pop-ups by `ppm_request_permission()` and `ppm_request_permissions()` are launched as `group mode` with the caller apps.   
> If the pop-up is terminated without full response, all the remaining requests that aren't responded by the user will be interpreted as 'Deny' action on behalf of the user. In this case, the app gets `PRIVACY_PRIVILEGE_MANAGER_REQUEST_RESULT_DENY_ONCE` response to those permissions. The app can again request for those **automatically denied** permissions anytime it needs.  
> User can use `ppm_request_permissions()` to request for multiple privileges instead of calling `ppm_request_permission()` multiple times.


@yl33 I slightly modified the sentence for more clarity, request you to verify the sentence technically and if you feel this is fine, kindly incorporate the changes

yl33 replied:
> User can use `ppm_request_permissions()` to request for multiple privileges instead of calling `ppm_request_permission()` multiple times.
This note is not for the users but for the developers. Modified your comment slightly.
Modified other phrases as you mentioned.

nawab-samsung commented on docs/application/web/guides/security/privacy-related-permissions.md:
```diff
@@ -18,6 +18,11 @@ The main features of the Privacy Privilege API include:
 
 For a list of privacy-related privileges, see [Security and API Privileges](../../tutorials/sec-privileges.md).
 
+> [!NOTE]
```
@yl33  Kindly refer to our [custom style guide](https://github.com/Samsung/tizen-docs/blob/master/styleguide/custom-style.md) to know more about new note styling.

We do not need an empty space after note tag, so:
> [!NOTE]
> Since Tizen 5.5, pop-ups by `requestPermission()` and `requestPermissions()` are launched as `group mode` with the caller apps.   
> If the pop-up is terminated without full response, all the remaining requests that aren't responded by the user will be interpreted as 'Deny' action on behalf of the user. In this case, the app gets `PPM_DENY_ONCE` response to those permissions. The app can again request for those **automatically denied** permissions anytime it needs.  
> User can use `requestPermissions()` to request for multiple privileges instead of calling `requestPermission()` multiple times.


@yl33 I slightly modified the sentence for more clarity, request you to verify the sentence technically and if you feel this is fine, kindly incorporate the changes

yl33 replied:
> User can use `requestPermissions()` to request for multiple privileges instead of calling `requestPermission()` multiple times.

This note is not for the users but for the developers. Modified your comment slightly.
Modified other phrases as you mentioned.

nawab-samsung commented on docs/application/web/guides/security/privacy-related-permissions.md:
```diff
@@ -18,6 +18,11 @@ The main features of the Privacy Privilege API include:
+>
+> Since Tizen 5.5, pop-ups by `requestPermission()` and `requestPermissions()` are launched as `group mode` with the caller apps.   
+> If the pop-up is terminated without full responses, all remained requests that aren't responded by the user will be treated as if the user selected 'Deny' for the requests. In this case, the app will get `PPM_DENY_ONCE` for those permissions and the app can request the **automatically denied** permissions again anytime it wants.  
+> If you need to request multiple privileges then use `requestPermissions()` instead of calling `requestPermission()` multiple times.
```
@yl33 
I see there is difference in method camel case notation style. 
`requestPermission()` method starts with small **r** , while as  file docs/application/dotnet/guides/security/privacy-related-permissions.md, it starts with capital **R**. `RequestPermission()`
Are they different? should'nt we follow consistent camel case notation styling for naming methods, incase they are same? 
If my question is not valid, kindly ignore this, else, kindly do the needful.

yl33 replied:
They are following naming rule of each language(web and dotnet). Don't need to fix.

nawab-samsung replied:
@yl33 thanks for clearng my doubt.

nawab-samsung commented on docs/application/dotnet/guides/security/privacy-related-permissions.md:
```diff
@@ -17,6 +17,12 @@ The main features of the `Tizen.Security.PrivacyPrivilegeManager` class include:
 For a list of privacy-related privileges, see [Security and API Privileges](../../tutorials/sec-privileges.md).
 
+> [!NOTE]
+>
```
Global: No space after [!Note]

Wrong usage 

\> [!NOTE]
\>
\> Example text 

Right usage 

\> [!NOTE]
\> Example text....

For more information on note styling, kindly refer to: 
https://github.com/Samsung/tizen-docs/blob/master/styleguide/custom-style.md#alerts-note-tip-important-caution-warning

yl33 replied:
Done

nawab-samsung commented:
@yl33 A gentle reminder!
I request you to kindly incorporate the comments.

yl33 commented on docs/application/dotnet/tutorials/sec-privileges.md:
```diff
@@ -34,10 +34,10 @@ application. For the application to use the API, the privilege must be
 switched it **on**.
 
-> **Note**  
+> [!NOTE]  
```
Done

yl33 commented on docs/application/dotnet/guides/security/privacy-related-permissions.md:
```diff
@@ -19,9 +19,9 @@ For a list of privacy-related privileges, see [Security and API Privileges](../.
-> If the pop-up is terminated without full responses, all remained requests that aren't responded by the user will be treated as if the user answered 'Deny' for the requests. In this case, the app will get `RequestResult.DenyOnce` for those permissions and the app can request the **automatically denied** permissions again anytime it wants.  
-> If you need to request multiple privileges then use `RequestPermissions()` instead of calling `RequestPermission()` multiple times.
+> Since Tizen 5.5, pop-ups by `RequestPermission()` and `RequestPermissions()` are launched as `group mode` with the caller apps.
+> If the pop-up is terminated without full response, all the remaining requests that aren't responded by the user will be interpreted as 'Deny' action on behalf of the user. In this case, the app gets `RequestResult.DenyOnce` response to those permissions. The app can again request for those **automatically denied** permissions anytime it needs.
```
Done

yl33 commented on docs/application/native/guides/security/privacy-related-permissions.md:
```diff
@@ -19,9 +19,9 @@ For a list of privacy-related privileges, see [Security and API Privileges](../.
-> If the pop-up is terminated without full responses, all remained requests that aren't responded by the user will be treated as if the user answered 'Deny' for the requests. In this case, the app will get `PRIVACY_PRIVILEGE_MANAGER_REQUEST_RESULT_DENY_ONCE` for those permissions and the app can request the **automatically denied** permissions again anytime it wants.  
-> If you need to request multiple privileges then use `ppm_request_permissions()` instead of calling `ppm_request_permission()` multiple times.
+> Since Tizen 5.5, pop-ups by `ppm_request_permission()` and `ppm_request_permissions()` are launched as `group mode` with the caller apps.
+> If the pop-up is terminated without full response, all the remaining requests that aren't responded by the user will be interpreted as 'Deny' action on behalf of the user. In this case, the app gets `PRIVACY_PRIVILEGE_MANAGER_REQUEST_RESULT_DENY_ONCE` response to those permissions. The app can again request for those **automatically denied** permissions anytime it needs.
```
Done

yl33 commented on docs/application/web/guides/security/privacy-related-permissions.md:
```diff
@@ -20,9 +20,9 @@ For a list of privacy-related privileges, see [Security and API Privileges](../.
-> If the pop-up is terminated without full responses, all remained requests that aren't responded by the user will be treated as if the user selected 'Deny' for the requests. In this case, the app will get `PPM_DENY_ONCE` for those permissions and the app can request the **automatically denied** permissions again anytime it wants.  
-> If you need to request multiple privileges then use `requestPermissions()` instead of calling `requestPermission()` multiple times.
+> Since Tizen 5.5, pop-ups by `requestPermission()` and `requestPermissions()` are launched as `group mode` with the caller apps.
+> If the pop-up is terminated without full response, all the remaining requests that aren't responded by the user will be interpreted as 'Deny' action on behalf of the user. In this case, the app gets `PPM_DENY_ONCE` response to those permissions. The app can again request for those **automatically denied** permissions anytime it needs.
```
Done

yl33 commented:
Updated as you reviewed and added details of the case that pop-up launched as `group mode` cause appfw confirmed that `group mode` launching is impossible for the apps with the other component type.

nawab-samsung commented on docs/application/web/guides/security/privacy-related-permissions.md:
```diff
@@ -18,6 +18,10 @@ The main features of the Privacy Privilege API include:
 For a list of privacy-related privileges, see [Security and API Privileges](../../tutorials/sec-privileges.md).
 
+> [!NOTE]
+> Since Tizen 5.5, if the caller app's `App component type` is `UI app` or `Component based app` with `Frame component` then pop-ups by `requestPermission()` and `requestPermissions()` are launched as `group mode` with the caller apps.
```
@yl33 
I have doubt here: 
1. Is it necessary to keep `App component type` is `UI app` or `Component based app` in code tags?
Kindly clarify the same.

Just in my view, can we write it like this: 
Since Tizen 5.5, if the caller app component type is UI app or component based app with frame component, then,  the pop-ups by `requestPermission()` and `requestPermissions()` are launched as `group mode` with the caller apps.

- This is just a suggestion, can you please give a better rephrase if you don't agree to my view point.
- Also verify for the technical meaning too, in case you agree with my suggestion.

yl33 replied:
Removed code tags.

nawab-samsung commented on docs/application/native/guides/security/privacy-related-permissions.md:
```diff
@@ -17,6 +17,11 @@ The main features of the Privacy Privilege Manager API include:
 For a list of privacy-related privileges, see [Security and API Privileges](../../tutorials/details/sec-privileges.md).
 
+> [!NOTE]
+> Since Tizen 5.5, if the caller app's `App component type` is `UI app` or `Component based app` with `Frame component` then pop-ups by `ppm_request_permission()` and `ppm_request_permissions()` are launched as `group mode` with the caller app.
```
@yl33
I have doubt here: 
1. Is it necessary to keep `App component type` is `UI app` or `Component based app` in code tags?
Kindly clarify the same.

Just in my view, can we write it like this: 
Since Tizen 5.5, if the caller app component type is UI app or component based app with frame component, then, the pop-ups by `requestPermission()` and `requestPermissions()` are launched as `group mode` with the caller apps.

- This is just a suggestion, can you please give a better rephrase if you don't agree to my view point.
- Also verify for the technical meaning too, in case you agree with my suggestion.

yl33 replied:
Removed code tags.

nawab-samsung commented on docs/application/dotnet/guides/security/privacy-related-permissions.md:
```diff
@@ -17,6 +17,11 @@ The main features of the `Tizen.Security.PrivacyPrivilegeManager` class include:
 For a list of privacy-related privileges, see [Security and API Privileges](../../tutorials/sec-privileges.md).
 
+> [!NOTE]
+> Since Tizen 5.5, if the caller app's `App component type` is `UI app` or `Component based app` with `Frame component` then pop-ups by `RequestPermission()` and `RequestPermissions()` are launched as `group mode` with the caller app.
```
@yl33
I have doubt here: 
1. Is it necessary to keep `App component type` is `UI app` or `Component based app` in code tags?
Kindly clarify the same.

Just in my view, can we write it like this: 
Since Tizen 5.5, if the caller app component type is UI app or component based app with frame component, then, the pop-ups by `RequestPermission()` and `RequestPermissions()` are launched as `group mode` with the caller apps.

- This is just a suggestion, can you please give a better rephrase if you don't agree to my view point.
- Also verify for the technical meaning too, in case you agree with my suggestion.

yl33 replied:
Removed code tags.

nawab-samsung commented on docs/application/dotnet/guides/security/privacy-related-permissions.md:
```diff
@@ -227,7 +232,7 @@ To check whether an application has permission to use a privilege, and to reques
 
-> **Note**
+> [!NOTE]
 >
```
@yl33  
Request you to check the following: 

Global, 
In whole file, can we remove extra space between: 
> [!NOTE]
> text....

yl33 replied:
Removed empty line after note tag.

nawab-samsung commented:
@yl33  gentle reminder, can we incorporate the said comments?

nawab-samsung commented on docs/application/dotnet/guides/security/privacy-related-permissions.md:
```diff
@@ -31,7 +36,7 @@ To enable your application to use the privacy-related permissions functionality:
     It means that the methods can be employed in any UI event handler (such as button click, timer event, system event, and application state change event). If you want to resolve privileges during application startup, call these methods from the Xamarin.Forms resume and start life-cycle methods (`Xamarin.Forms.Application.OnResume()` and `Xamarin.Forms.Application.OnStart()`).
 
-    > **Note**
+    > [!NOTE]
     >
     > The `Tizen.Security.PrivacyPrivilegeManager` class is not thread-safe.
```
@yl33 I guess you might have missed editing note styling here.

yl33 replied:
Right. Removed that one, too.
